lsp/handlers/lsp_router.cpp

```diff
@@ -0,0 +1,138 @@
+#include "lsp_router.h"
+
+#include "initialize.hpp"
+
+#include "common/log/log.h"
+
+#include "protocol/error_codes.h"
+#include "text_document/document_symbol.hpp"
+#include "text_document/document_synchronization.hpp"
+#include "text_document/go_to.hpp"
+#include "text_document/hover.hpp"
+
+#include "third-party/fmt/core.h"
+
+LSPRoute::LSPRoute() : m_route_type(LSPRouteType::NOOP) {}
+
+LSPRoute::LSPRoute(std::function<void(Workspace&, json)> notification_handler)
+    : m_route_type(LSPRouteType::NOTIFICATION), m_notification_handler(notification_handler) {}
+
+LSPRoute::LSPRoute(std::function<void(Workspace&, json)> notification_handler,
+                   std::function<std::optional<json>(Workspace&, json)> post_notification_publish)
+    : m_route_type(LSPRouteType::NOTIFICATION),
+      m_notification_handler(notification_handler),
+      m_post_notification_publish(post_notification_publish) {}
+
+LSPRoute::LSPRoute(std::function<std::optional<json>(Workspace&, int, json)> request_handler)
+    : m_route_type(LSPRouteType::REQUEST_RESPONSE), m_request_handler(request_handler) {}
+
+void LSPRouter::init_routes() {
+  m_routes["initialize"] = LSPRoute(initialize_handler);
+  m_routes["initialize"].m_generic_post_action = [](Workspace& workspace) {
+    workspace.set_initialized(true);
+  };
+  m_routes["initialized"] = LSPRoute();
+  m_routes["textDocument/documentSymbol"] = LSPRoute(document_symbols_handler);
+  m_routes["textDocument/didOpen"] = LSPRoute(did_open_handler, did_open_push_diagnostics);
+  m_routes["textDocument/didChange"] = LSPRoute(did_change_handler, did_change_push_diagnostics);
+  m_routes["textDocument/didClose"] = LSPRoute(did_close_handler);
+  m_routes["textDocument/hover"] = LSPRoute(hover_handler);
+  m_routes["textDocument/definition"] = LSPRoute(go_to_definition_handler);
+}
+
+json error_resp(ErrorCodes error_code, std::string error_message) {
+  json error{
+      {"code", static_cast<int>(error_code)},
+      {"message", error_message},
+  };
+  return json{{"error", error}};
+}
+
+std::string LSPRouter::make_response(const json& result) {
+  json content = result;
+  content["jsonrpc"] = "2.0";
+
+  std::string header;
+  header.append("Content-Length: " + std::to_string(content.dump().size()) + "\r\n");
+  header.append("Content-Type: application/vscode-jsonrpc;charset=utf-8\r\n");
+  header.append("\r\n");
+  return header + content.dump();
+}
+
+std::optional<std::vector<std::string>> LSPRouter::route_message(
+    const MessageBuffer& message_buffer,
+    AppState& appstate) {
+  json body = message_buffer.body();
+  auto method = body["method"];
+  lg::info(method);
```

Comment:
Does this work? Won't the type of `method` be `json` here?

Reply:
Huh, hadn't noticed that because it indeed does work 
![image](https://user-images.githubusercontent.com/13153231/179439186-f4a73690-cee6-45bc-8305-df4dffbcf133.png)
nlohman::json must define some custom fmt::formatters

Reply:
ah, nice!